### DISCUSSION TITLE
Audio Stream Send Problem

### DISCUSSION THREAD

**Discussion by mkemalgokce** [paullouisageneau/libdatachannel#719] [Q&A]
### Hello, 

I want to create video call application between libdatachannel c++ client and web client. So i created signal server, turn server and stun server. I want to send my video and audio datas with multiple tracks.(Video track and audio track). My libdatachannel client sends offer to web client. And my connection completes. And video stream starts succesfully (i can receive videos from both sides).
**But i have problems on audio side.** Web client sends audio and i can receive web audio from libdatachannel client. But i can't receive libdatachannel audio from web client. This is my problem. I can't resolve this problem.

I used gstreamer for sending audio stream.( I choose gstreamer because your examples like that).

**This is my gstreamer pipeline :**
`gst-launch-1.0 autoaudiosrc ! audioconvert ! audioresample ! queue ! opusenc ! audio/x-opus, profile=constrained-baseline ! rtpopuspay pt=96 mtu=1200 ! udpsink host=127.0.0.1 port=5002`

**These are libdatachannel codes :**
```
        rtc::Description::Audio audio("audio", rtc::Description::Direction::SendRecv);

        const rtc::SSRC ssrc2 = 42;
        audio.addSSRC(ssrc2, "audio", "stream1", "audio");

        audio.addOpusCodec(111);

        shared_ptr<rtc::Track> audioTracker = pc->addTrack(audio);

        auto audioSession = std::make_shared<rtc::RtcpReceivingSession>();



        audioTracker->setMediaHandler(audioSession);


        SOCKET audio_sock_receive = socket(AF_INET, SOCK_DGRAM, 0);
        sockaddr_in audio_addr_receive = {};
        audio_addr_receive.sin_family = AF_INET;
        audio_addr_receive.sin_addr.s_addr = inet_addr("127.0.0.1");
        audio_addr_receive.sin_port = htons(5001);

        audioTracker->onMessage([audioSession, audio_sock_receive, audio_addr_receive](rtc::binary message) {
            std::cout << "Audio Come! \n";
            sendto(audio_sock_receive, reinterpret_cast<const char *>(message.data()), int(message.size()), 0,
            reinterpret_cast<const struct sockaddr *>(&audio_addr_receive), sizeof(audio_addr_receive));
        }, nullptr);
```

    SOCKET sock_send = socket(AF_INET, SOCK_DGRAM, 0);
    sockaddr_in addr_send = {};
    addr_send.sin_family = AF_INET;
    addr_send.sin_addr.s_addr = inet_addr("127.0.0.1");
    addr_send.sin_port = htons(5002);
    if (bind(sock_send, reinterpret_cast<const sockaddr *>(&addr_send), sizeof(addr_send)) < 0)
             throw std::runtime_error("Failed to bind UDP socket on 127.0.0.1:5002");

    int rcvBufSize = 212992;

    setsockopt(sock_send, SOL_SOCKET, SO_RCVBUF, reinterpret_cast<const char *>(&rcvBufSize),
               sizeof(rcvBufSize));

    char buffer[BUFFER_SIZE];
    int len;
    while ((len = recv(sock_send, buffer, BUFFER_SIZE, 0)) >= 0) {

        if (len < sizeof(rtc::RtpHeader) || !audioTracker->isOpen())
            continue;

        auto rtp = reinterpret_cast<rtc::RtpHeader *>(buffer);
        rtp->setSsrc(ssrc);
        std::cout << "Audio Sent! \n";

        audioTracker->send(reinterpret_cast<const std::byte *>(buffer), len);

    }

**And also i noticed payload types. I used different payload types for video and audio**

I need some help. Thanks a lot!

Have a nice day.

**Comment by paullouisageneau** [comment]
It seems the gstreamer pipeline uses payload type 96 while you set payload type 111 in the description, therefore the web browser may discard the data as it expects payload type 111.

**Reply by mkemalgokce** [reply]
Thanks for your help. It helps me a lot. I am really pleased.😀
